### PR TITLE
Rework of client connect and disconnect behaviour

### PR DESCRIPTION
**Pull request by rodja**
As a result of discussion in https://github.com/zauberzeug/nicegui/discussions/232 this pull request changes the behaviour of client connections in the following ways:

- new methods `client.on_connected` and `client.on_disconnected` to register callbacks for events of a specific client
- renaming `await client.handshake()` to `await client.connected()` for more consitency
- new coroutine `await client.disconnected()` to block execution in a page builder method until connection is closed
- `ui.on_connect` and `ui.on_disconnect` are now always called if any client connects (before it had the behaviour of the newly introduced `client.on_connected` and `client.on_disconnected` which was unintuitiv.

ToDos

- [x] fix tests
- [x] change behaviour of `ui.on_connect` and `ui.on_disconnect` (including their usage)
- [x] fix test for Python 3.11
- [x] add doc strings for `ui.on_connect` and `ui.on_disconnect`

### CODE REVIEW

**Comment by rodja** [comment]
I think it would be good to have this reviewed. Even if I'm not 100% sure about the `client` parameter in `ui.on_connect` and `ui.on_disconnect`...

File: nicegui/helpers.py [hunk 1]
```diff
@@ -14,7 +14,7 @@ def is_coroutine(object: Any) -> bool:
     return asyncio.iscoroutinefunction(object)
 
 
-def safe_invoke(func: Union[Callable, Awaitable], client: Optional[Client] = None) -> None:
+def safe_invoke(func: Union[Callable, Awaitable], client: Optional[Client] = None, *args: List[Any]) -> None:
```
Comment (falkoschindler) [inline nicegui/helpers.py:17]: "For variable positional arguments (*args) and variable keyword arguments (**kw) you only need to specify the expected value for one such argument."
https://stackoverflow.com/a/37032111/3419103
https://peps.python.org/pep-0484/#arbitrary-argument-lists-and-default-argument-values

File: tests/test_lifecycle.py [hunk 2]
```diff
@@ -1,21 +1,52 @@
...
     screen.should_contain('42')
+
+
+def test_connect_disconnect_is_called_for_each_client(screen: Screen):
+    events: List[str] = []
+
+    @ui.page('/')
+    def page(client: Client):
+        ui.label(f'client id: {client.id}')
+    ui.on_connect(lambda c: events.append(f'|connect {c.id}|'))
```
Comment (falkoschindler) [inline tests/test_lifecycle.py:40]: I guess the pipe used to be a separator but is not needed anymore.

File: tests/test_page.py [hunk 3]
```diff
@@ -114,18 +114,36 @@ async def takes_a_while() -> None:
...
     screen.should_contain('delayed data has been loaded')
 
 
-def test_adding_elements_after_handshake(screen: Screen):
+def test_wait_for_disconnect(screen: Screen):
+    events = []
+
+    @ui.page('/')
+    async def page(client: Client):
+        nonlocal events
```
Comment (falkoschindler) [inline tests/test_page.py:129]: `nonlocal` declaration is not needed.

**Comment by falkoschindler** [comment]
Although this approach looks promising, I'm not sure if we really should squeeze it into the 1.1 release. There are already 5 closed issues waiting for release and I could imagine there are going to be some more iterations for the new lifecycle API.

Regarding the new API:

- Renaming "handshake" into "connected" sounds reasonable.
- A new coroutine `client.disconnected()` seems to be a handy counterpart to `client.connected()`.
- `client.on_connected` and `client.on_disconnected`: This is a bit redundant, but probably more common and thus more intuitive than the coroutines.
- `ui.on_connect` and `ui.on_disconnect` now being always called: This worries me a bit. NiceGUI *elements* are automatically placed in the current (auto-)context on the current page. And (almost) all our *functions* `add_body_html`, `add_head_html`, `run_javascript`, `notify`, `open` and `timer` impact the page/client/context they are called within. Exceptions are `on_startup`, `on_shutdown`, `shutdown`, `add_static_files`, which make only sense in a global context, and `update`, which gets the list of elements explicitly. The new `on_connect` and `on_disconnect` could refer to a single client like the other functions, or act globally like the other lifecycle hooks. So I wonder how the user should notice what impact a NiceGUI command has.

**Comment by rodja** [comment]
> Although this approach looks promising, I'm not sure if we really should squeeze it into the 1.1 release. There are already 5 closed issues waiting for release and I could imagine there are going to be some more iterations for the new lifecycle API.

You are possibly right. But I would like to not release a 1.2 in a few days. Let's try to figure this out. If we are not ready to merge on Monday we can still release 1.1 without a new lifecycle API.

File: nicegui/helpers.py (hunk 1 above)
Comment (rodja) [reply nicegui/helpers.py:17]: Thanks! I changed the code a bit to make the client parameter optional.

File: tests/test_lifecycle.py (hunk 2 above)
Comment (rodja) [reply tests/test_lifecycle.py:40]: True

File: website/reference.py [hunk 4]
```diff
@@ -486,22 +486,23 @@ def colors_example():
...
 
-        async def countdown():
-            for i in [5, 4, 3, 2, 1, 0]:
-                l.text = f'{i}...' if i else 'Take-off!'
-                await asyncio.sleep(1)
+        async def decrement(client: Client):
+            counter.value -= 1
 
-        ui.on_connect(countdown)
+        counter = ui.number('connections', value=0).props('readonly').classes('w-24')
```
[the remark this answers is not in the document]
Comment (rodja) [reply website/reference.py]: Yes. That was crude. I've created another one...

**Comment by rodja** [comment]
> ui.on_connect and ui.on_disconnect now being always called: This worries me a bit. NiceGUI elements are automatically placed in the current (auto-)context on the current page. And (almost) all our functions add_body_html, add_head_html, run_javascript, notify, open and timer impact the page/client/context they are called within. Exceptions are on_startup, on_shutdown, shutdown, add_static_files, which make only sense in a global context, and update, which gets the list of elements explicitly. The new on_connect and on_disconnect could refer to a single client like the other functions, or act globally like the other lifecycle hooks.

We now have `client.on_connect` and `client.on_disconnect` for context-awareness. Therefore we could simply remove the old lifecycle functions `ui.on_connect` and `ui.on_disconnect`.

If we keep them (or re-add them later), they should act globally. I think it would be more consistent if all lifecycle events are global. The deviation was possibly the root cause of the misunderstanding in https://github.com/zauberzeug/nicegui/discussions/232#discussioncomment-4615426.

> So I wonder how the user should notice what impact a NiceGUI command has.

Thats a good point. We could move the functions with global impact into `app`. Then it would be more clear that these functions act on the whole backend and not just the current client.

**Comment by rodja** [comment]
> We could move the functions with global impact into app. Then it would be more clear that these functions act on the whole backend and not just the current client.

I've another suggestion: We could put the globally impacting lifecycle functions into a submodule. For example `ui.lifecycle.on_connect()`. This way it's more clear that this is not a normal `ui` call with auto-context behaviour.
I already pushed an implementation to this pull-request. 

A similar namespace trick could be used for routing: `ui.routing. add_static_files()`.

**Comment by rodja** [comment]
We decided to move the global affecting functions into `app`.

**Pull request merged**